File: 01-arrays_&_hashing/00128-Longest_Consecutive_Sequence/00128-longest_consecutive_sequence.go
// 00128: Longest Consecutive Sequence
// https://leetcode.com/problems/longest-consecutive-sequence/

package main

import (
	"fmt"
	"math"
)

// SOLUTION
func longestConsecutive(nums []int) int {
	result := 0
	tmap := make(map[int]int)

	for _, i := range nums {
		if tmap[i]!=0 {continue}
		
		left := tmap[i-1]
		right := tmap[i+1]
		sum := left + right + 1
		result = int(math.Max(float64(result), float64(sum)))

		if left>0 {tmap[i-left] = sum}
		if right>0 {tmap[i+right] = sum}
		tmap[i] = sum		
	}

	return result
}


func main() {
	// INPUT
	nums := []int{100,4,200,1,3,2}

	// OUTPUT
	result := longestConsecutive(nums)
	fmt.Println(result)
}

File: 01-arrays_&_hashing/00128-Longest_Consecutive_Sequence/00128-longest_consecutive_sequence.cpp
// 00128: Longest Consecutive Sequence
// https://leetcode.com/problems/longest-consecutive-sequence/

#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    // SOLUTION
    int longestConsecutive(vector<int>& nums) {
        int result = 0;
        unordered_map<int,int> tmap;

        for (int i : nums) {
            if (tmap[i]) continue;
            result = max(
                            result, 
                            tmap[i]
                                = tmap[i + tmap[i+1]]
                                = tmap[i - tmap[i-1]]
                                = tmap[i+1] + tmap[i-1] + 1
                        );
        }

        return result;
    }
};

int main() {
    Solution o;

    // INPUT
    vector<int> nums = {100,4,200,1,3,2};

    // OUTPUT
    auto result = o.longestConsecutive(nums);
    cout<<result<<endl;

    return 0;
}